include/mbgl/storage/default_file_source.hpp

```diff
@@ -17,6 +17,7 @@ template <typename T> class Thread;
 } // namespace util
 
 class ResourceTransform;
+using PathChangeCallback = std::function<void ()>;
```

Comment:
Should there be a way to communicate failure?

Right now, `void OfflineDatabase::changePath(const std::string&)` can throw exception, should developer be notified when `setResourceCachePath` fails?

Reply:
What's the way for `void OfflineDatabase::changePath(const std::string&)` to throw? Is that only the `assert(!db)` and `assert(statements.empty())` assertions in `OfflineDatabase::initialize()`? If that's the case, we are logging a failure to cleanup beforehand, so I don't think we need to send failure callback just before we crash anyway. Same if we fail to open the database.

Reply:
I think the underlying problem here is that we don't have a proper mechanism for detecting errors when we failed to open the database/

Reply:
Is this something that we'd be able to follow up on in a separate PR?

Reply:
@LukasPaczos  almost every line in `OfflineDatabase::initialize` throws an exception (db->*, file deletion). `ResourcesCachePathChangeCallback` already provides `onError`, that might be used on SDK side to notify developer when changePath failed.

This PR did not introduce that issue, so, I think fix for propagating an error on `initialize` can be done in separate PR. May be worth creating issue + adding TODO note for `PathChangeCallback` that refers to an issue.

@tmpsantos wdyt?

Reply:
Created https://github.com/mapbox/mapbox-gl-native/issues/14759 and added a `TODO`.